src/strings/ops.c

```diff
@@ -206,38 +205,130 @@ static void turn_32bit_into_8bit_unchecked(MVMThreadContext *tc, MVMString *str)
 
     MVM_free(old_buf);
 }
+/* Checks if the next num_graphs graphemes in the iterator can fit into 8 bits.
+ * This was written to take advantage of SIMD vectorization, so we use a multiple
+ * bitwise operations to check, and biwise OR it with val. Care must be taken
+ * to not use any variables altered by the loop outside of the loop and to not
+ * have any branching or funcion calls. `i` is not used outside the loop
+ * `val` is allowed as biwise OR works with the vectorization well.
+ * NOTE: Final state of the grapheme iterator passed into i is undefined. It should
+ * not be reused. */
+static int string_can_be_8bit(MVMThreadContext *tc, MVMGraphemeIter *gi, MVMStringIndex num_graphs) {
+    int val = 0;
+    MVMStringIndex pos = 0;
+    while (1) {
+        size_t strand_len = gi->end - gi->pos;
```

Comment:
I think I'd maybe prefer that we introduce static inline functions for dealing with the grapheme iterator data structure, rather than poking in to it. The design so far was - at least hopefully - such that we didn't poke into the struct directly, but worked with it in terms of static inline functions. Since they're effectively just a nicer way to do macros, they shouldn't have any performance impact, while keeping it easier to change things in the future.